File: 2021/Day 11/TS/part2.ts
import fs from 'fs'
import path from 'path'

const file: string = fs.readFileSync(path.resolve(__dirname, '../data.txt'), 'utf-8')
const octopi: number[][] = file.split('\n').map((line) => line.split('').map((char) => parseInt(char)))

const delta = [
    [-1, -1],
    [0, -1],
    [1, -1],
    [-1, 0],
    [1, 0],
    [-1, 1],
    [0, 1],
    [1, 1],
]

let flashes = 0
let steps = 0
while (flashes != octopi.length * octopi[0].length) {
    incrementAll(octopi)
    flashes = flash(octopi)
    steps++
}

// for (let line of octopi) console.log(JSON.stringify(line).replace(/,/g, ' '))
console.log('Steps:', steps)

// Increment all
function incrementAll(octopi: number[][]) {
    for (let line of octopi) {
        for (let x = 0; x < line.length; x++) line[x]++
    }
}

// Flash
function flash(octopi: number[][], max = 9): number {
    let flashes = 0
    // Initialize queue
    for (let y = 0; y < octopi.length; y++) {
        for (let x = 0; x < octopi[y].length; x++) {
            if (octopi[y][x] > max) flashes += flashOctopus(octopi, x, y, max)
        }
    }
    return flashes
}

// Flash a single octopus
function flashOctopus(octopi: number[][], x: number, y: number, max: number): number {
    let flashes = 1
    octopi[y][x] = 0 // Flash self
    // Flash neighbours
    for (let [dx, dy] of delta) {
        if (!octopi[y + dy] || !octopi[y + dy][x + dx]) continue // Neighbour doesn't exists
        if (octopi[y + dy][x + dx] == 0) continue // Neighbour already flashed
        octopi[y + dy][x + dx]++ // Increment neighbour
        if (octopi[y + dy][x + dx] > max) flashes += flashOctopus(octopi, x + dx, y + dy, max)
    }
    return flashes
}

File: 2021/Day 11/C++/part2.cpp
#include <fstream>
#include <iostream>

// Namespaces
using namespace std;  // Bad practice, I know

// Constants
const int delta[8][2] = {{-1, -1}, {0, -1}, {1, -1}, {-1, 0}, {1, 0}, {-1, 1}, {0, 1}, {1, 1}};

// Functions
// Flash a single octopus
int flashOctopus(int octopi[10][10], int x, int y) {
    int flashes = 1;
    octopi[y][x] = 0;  // Flash self
    // Flash neighbours
    for (auto d : delta) {
        int x2 = x + d[0], y2 = y + d[1];
        if (x2 < 0 || x2 > 9 || y2 < 0 || y2 > 9) continue;  // Neighbor doesn't exist
        if (octopi[y2][x2] == 0) continue;                   // Neighbour already flashed
        octopi[y2][x2]++;                                    // Increment neighbour
        if (octopi[y2][x2] > 9) flashes += flashOctopus(octopi, x2, y2);
    }
    return flashes;
}

// Flash the octopi like xmas lights
int flash(int octopi[10][10]) {
    int flashes = 0;
    // Flash some octopi
    for (int i = 0; i < 100; i++) {
        if (octopi[i / 10][i % 10] > 9) flashes += flashOctopus(octopi, i % 10, i / 10);
    }
    return flashes;
}

// Main
int main() {
    // Get input
    ifstream file("../data.txt");
    int octopi[10][10];
    char number;
    int i = 0;
    while (file >> number) {
        octopi[i / 10][i % 10] = number - 48;
        i++;
    }

    // Count flashes
    int flashes = 0;
    int steps = 0;
    while (flashes != 100) {
        for (int i = 0; i < 100; i++) octopi[i / 10][i % 10]++;  // increment all
        flashes = flash(octopi);
        steps++;
    }

    // Output result
    cout << "Steps: " << steps << endl;

    return 0;
}